Doc/library/typing.rst

```diff
@@ -585,6 +587,31 @@ These can be used as types in annotations and do not support ``[]``.
    .. versionadded:: 3.5.4
    .. versionadded:: 3.6.2
 
+.. data:: Self
+
+   Special type to represent the current enclosed class.
+   For example::
+
+      from typing import Self
+
+      class Foo:
+         def returns_self(self) -> Self:
+            ...
+            return self
+
+   .. note::
+
+      This code would be semantically equivalent to a bound :class:`TypeVar` with the `bound=Foo`.
+
+   This is especially useful for:
+
+      - :class:`classmethod`\s that are used as alternative constructors
+      - Annotating an :meth:`object.__enter__` method which returns self
+
+   For more info see :pep:`673`.
```

Comment:
```suggestion

   This annotation is semantically equivalent to using a :class:`TypeVar` with `bound=Foo` as both the return annotation and the annotation for the `self` parameter.

   Common use cases include:

      - :class:`classmethod`\s that are used as alternative constructors
      - Annotating an :meth:`object.__enter__` method which returns self

   For more information, see :pep:`673`.
```

Reply:
Sorry this might need some manual adjustment (newlines and line length). Apart from that I'm happy with the change though! We should be able to get it merged once main is unblocked.